packages/mui-system/src/style.js

```diff
@@ -39,6 +39,14 @@ export function getStyleValue(themeMapping, transform, propValueFinal, userValue
     value = transform(value, userValue, themeMapping);
   }
 
+  if (typeof value !== 'string' && typeof value !== 'number' && typeof value !== 'boolean') {
```

Comment:
this is what I suggest -

```js
if (typeof value === 'object') {
  if (process.env.NODE_ENV !== 'production') {
    console.warn(
      `MUI: Invalid value found in theme for prop: "${propValueFinal}". It should be a string or number. Check if you forgot to add the correct dotted notation, eg, "background.paper" instead of "background".`,
    );
  }
  value = userValue;
}
```

and move the check before the call to `transform`, just after line 37 as transform can return nested style object that translate into correct CSS.

And add similar test in `style.test.js` as well as `Typography.test.ts`